zk/conn.go

```diff
@@ -266,13 +279,36 @@ func WithEventCallback(cb EventCallback) connOption {
 	}
 }
 
+func (c *Conn) closeWithError(err error) {
+	c.closeErrMu.Lock()
+	if c.closeErr != nil {
+		// Conn is already closed.
+		c.closeErrMu.Unlock()
+		return
+	}
+	c.closeErr = err
+	c.closeErrMu.Unlock()
+
+	close(c.sendChan)
+	c.flushUnsentRequests(err)
+}
+
 func (c *Conn) Close() {
 	close(c.shouldQuit)
 
+	respChan := make(chan struct{}, 1)
+
+	go func() {
+		_, _ = c.request(opClose, &closeRequest{}, &closeResponse{}, nil)
```

Comment:
I don't understand why you are calling request, rather than queueRequest. If there is a reason, it needs a comment.